File: ruby/p12.rb
#!/usr/bin/ruby

#What is the value of the first triangle number to have over five hundred divisors

def numDivisors(x)
	sqrtx = Math.sqrt(x)
	i = 1
	numDiv = 0
	divisorList = []
	while i <= sqrtx
		if x%i==0
			numDiv+=(i==sqrtx)?1:2
			divisorList.push(x/i,i)
		end
		i+=1
	end
	puts divisorList.uniq.sort.join(',')
	return numDiv
end

for i in 2..20
	puts i.to_s + ": " + numDivisors(i).to_s
	puts
end
exit

triNum = 1
triAdder = 2

while numDivisors(triNum) < 500
	triNum+=triAdder
	triAdder+=1
end

puts triNum

File: c++/p12.cpp
#include "p12.hpp"
#include <math.h>

MAKE_PROTOTYPE(p12)

//Let us list the factors of the first seven triangle numbers:
//
// 1: 1
// 3: 1,3
// 6: 1,2,3,6
// 10: 1,2,5,10
// 15: 1,3,5,15
// 21: 1,3,7,21
// 28: 1,2,4,7,14,28
//
//We can see that 28 is the first triangle number to have over five divisors.
//
//What is the value of the first triangle number to have over five hundred divisors?

uint getNumDivisors(uint num)
{
  uint divisor = 1, numDivisors = 0;
  uint limit = (uint)sqrt(num); 

  while (divisor <= limit)
  {
    if ( num % divisor == 0 )
    {
      numDivisors += 2;
      if ( divisor == limit )
        numDivisors--;
    }
    ++divisor;
  }
  return numDivisors;
}



uint p12::execute() 
{
  uint trinum = 0, i = 1, numDivisors = 1;

  while (numDivisors < 500)
  {
    trinum += i++;
    numDivisors = getNumDivisors(trinum);
  }

  return trinum;
}
